modin/core/dataframe/pandas/metadata/dtypes.py

```diff
@@ -148,3 +150,141 @@ def __contains__(self, key):  # noqa: GL08
         if not self.is_materialized:
             self.get()
         return key in self._value
+
+
+class LazyProxyCategoricalDtype(pandas.CategoricalDtype):
+    """
+    A lazy proxy representing ``pandas.CategoricalDtype``.
+
+    Parameters
+    ----------
+    categories : list-like, optional
+    ordered : bool, default: False
+
+    Notes
+    -----
+    Important note! One shouldn't use the class' constructor to instantiate a proxy instance,
+    it's intended only for compatibility purposes! In order to create a new proxy instance
+    use the appropriate class method `._build_proxy(...)`.
+    """
+
+    def __init__(self, categories=None, ordered=False):
+        # These will be initialized later inside of the `._build_proxy()` method
+        self._parent, self._column_name, self._categories_val, self._materializer = (
+            None,
+            None,
+            None,
+            None,
+        )
+        super().__init__(categories, ordered)
+
+    def _update_proxy(self, parent, column_name):
+        """
+        Create a new proxy, if either parent or column name are different.
+
+        Parameters
+        ----------
+        parent : object
+            Source object to extract categories on demand.
+        column_name : str
+            Column name of the categorical column in the source object.
+
+        Returns
+        -------
+        pandas.CategoricalDtype or LazyProxyCategoricalDtype
+        """
+        if self._is_materialized:
+            # The parent has been materialized, we don't need a proxy anymore.
+            return pandas.CategoricalDtype(self.categories, ordered=self._ordered)
+        elif parent is self._parent and column_name == self._column_name:
+            return self
+        else:
+            return LazyProxyCategoricalDtype._build_proxy(
+                parent, column_name, self._materializer
+            )
+
+    @classmethod
+    def _build_proxy(cls, parent, column_name, materializer):
+        """
+        Construct a lazy proxy.
+
+        Parameters
+        ----------
+        parent : object
```

Comment:
Why is the `parent` type so generic? Would it be better to reduce the set of allowed types?